observers/qos_correction.cpp

```diff
@@ -0,0 +1,200 @@
+#include <list>
+#include <vector>
+
+#include "observers/qos_correction.hpp"
+
+#include "mesos/slave/oversubscription.hpp"
+
+#include "messages/serenity.hpp"
+
+#include "serenity/wid.hpp"
+
+namespace mesos {
+namespace serenity {
+
+using std::list;
+
+/**
+ * Checks contentions and choose executors to kill.
+ * Currently we have sorted Contentions by severity.
+ * We are iterating from the most serious contention to the less serious.
+ * WARNING(bplotka): This interpreter assumes that severity is reflecting
+ * how many resources is needed for PR job to not starve again.
+ */
+Try<QoSCorrections> ContentionInterpreters::severityBasedCpuContention(
+    Contentions currentContentions,
+    ResourceUsage currentUsage) {
+  // Product.
+  QoSCorrections corrections;
+
+  // BE tasks.
+  list<ResourceUsage_Executor> possibleAggressors =
+      filterPrExecutors(currentUsage);
+
+  // This interpreter controls CPU Contentions only.
+  // TODO(bplotka): Consider sorting them by CpuUsage
+  possibleAggressors.sort(QoSCorrectionObserver::compareCpuAllocated);
+
+  // To not rescue the same PR executor more then one time.
+  std::list<WID> correctedPrExecutors;
+  // We save here aggressors to be killed.
+  std::list<slave::QoSCorrection_Kill> aggressorsToKill;
+  float_t severity_balance = 0;
```

Comment:
When severity reflects how bad contention is and how many cpus we should "release" to solve the contention, if we kill executor in which usage > severity we save the difference as a balance for next contentions.